libs/ml/examples/distributed_learning/main.cpp

```diff
@@ -0,0 +1,420 @@
+    }
+
+    // Apply gradients to own model
+    {
+      std::lock_guard<std::mutex> l(model_mutex_);
+      g_.ApplyGradients(gradients);
+    }
+  }
+
+  /**
+   * Rewrites current model with given one
+   * @param new_weights Vector of weights that represent model
+   */
+  void SetWeights(TensorVectorType &new_weights)
+  {
+    std::lock_guard<std::mutex> l(const_cast<std::mutex &>(model_mutex_));
+    g_.set_weights(new_weights);
+  }
+
+private:
+  // Client's own graph and mutex to protect it's weights
+  fetch::ml::Graph<TensorType> g_;
+  std::mutex                   model_mutex_;
+
+  // Client's own dataloader
+  fetch::ml::dataloaders::MNISTLoader<TensorType, TensorType> dataloader_;
+
+  // Connection to other nodes
+  std::vector<std::shared_ptr<TrainingClient>> peers_;
+
+  // Access to coordinator
+  std::shared_ptr<Coordinator> coordinator_ptr_;
+
+  // Gradient queue access and mutex to protect it
+  std::queue<TensorVectorType> gradient_queue;
+  std::mutex                   queue_mutex_;
+
+  // random number generator for shuffling peers
+  fetch::random::LaggedFibonacciGenerator<> gen_;
+
+  // Learning hyperparameters
+  SizeType batch_size_      = BATCH_SIZE;
+  SizeType test_set_ratio_  = TEST_SET_RATIO;
+  SizeType number_of_peers_ = NUMBER_OF_PEERS;
+  SizeType learning_rate_   = LEARNING_RATE
+
+      // Client id (identification name)
+      std::string id_;
+
+  // Timestamp for logging
+  std::string GetTimeStamp()
+  {
+    // TODO: Implement timestamp
+    return "TIMESTAMP";
+  }
+};
+
+int main(int ac, char **av)
+{
+  if (ac < 3)
+  {
+    std::cout << "Usage : " << av[0]
+              << " PATH/TO/train-images-idx3-ubyte PATH/TO/train-labels-idx1-ubyte" << std::endl;
+    return 1;
+  }
+
+  std::shared_ptr<Coordinator> coordinator = std::make_shared<Coordinator>();
+
+  std::cout << "FETCH Distributed MNIST Demo -- Synchronised" << std::endl;
+  std::srand((unsigned int)std::time(nullptr));
+
+  std::vector<std::shared_ptr<TrainingClient>> clients(NUMBER_OF_CLIENTS);
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
+  {
+    // Instanciate NUMBER_OF_CLIENTS clients
+    clients[i] = std::make_shared<TrainingClient>(av[1], av[2], std::to_string(i));
+  }
+
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
+  {
+    // Give every client the full list of other clients
+    clients[i]->AddPeers(clients);
+
+    // Give each client pointer to coordinator
+    clients[i]->SetCoordinator(coordinator);
+  }
+
+  // Main loop
+  for (unsigned int it(0); it < NUMBER_OF_ITERATIONS; ++it)
+  {
+
+    // Start all clients
+    coordinator->state = CoordinatorState::RUN;
+    std::cout << "================= ITERATION : " << it << " =================" << std::endl;
+    std::list<std::thread> threads;
+    for (auto &c : clients)
+    {
+      threads.emplace_back([&c] { c->MainLoop(); });
+    }
+
+    std::this_thread::sleep_for(std::chrono::seconds(SYNCHRONIZATION_INTERVAL));
```

Comment:
add a boolean flag for synchronisation On/Off
At the moment if you want to try out fully asynchronous mode the only way to do it is to make this sleep forever.